File: codegen/output/csharp/Player.cs
using System;
using System.Runtime.InteropServices;


///
public class Player
{
  public IntPtr ptr;
  protected int ID;
  protected int iteration;

  public Player()
  {
  }

  public Player(IntPtr p)
  {
    ptr = p;
    ID = Client.playerGetId(ptr);
    iteration = BaseAI.iteration;
  }

  public bool validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.players.Length; i++)
    {
      if(BaseAI.players[i].ID == ID)
      {
        ptr = BaseAI.players[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }

    //commands

  ///Allows a player to display messages on the screen
  public bool talk(string message)
  {
    validify();
    return (Client.playerTalk(ptr, message) == 0) ? false : true;
  }

    //getters


  ///Unique Identifier
  public int Id
  {
    get
    {
      validify();
      int value = Client.playerGetId(ptr);
      return value;
    }
  }

  ///Player's Name
  public string PlayerName
  {
    get
    {
      validify();
      IntPtr value = Client.playerGetPlayerName(ptr);
      return Marshal.PtrToStringAuto(value);
    }
  }

  ///Time remaining, updated at start of turn
  public float Time
  {
    get
    {
      validify();
      float value = Client.playerGetTime(ptr);
      return value;
    }
  }

  ///The player's current reef health
  public int CurReefHealth
  {
    get
    {
      validify();
      int value = Client.playerGetCurReefHealth(ptr);
      return value;
    }
  }

  ///Currency for fish
  public int SandDollars
  {
    get
    {
      validify();
      int value = Client.playerGetSandDollars(ptr);
      return value;
    }
  }

}

File: client/cpp/Player.cpp
// -*-c++-*-

#include "Player.h"
#include "game.h"


Player::Player(_Player* pointer)
{
    ptr = (void*) pointer;
}

int Player::id()
{
  return ((_Player*)ptr)->id;
}

char* Player::playerName()
{
  return ((_Player*)ptr)->playerName;
}

float Player::time()
{
  return ((_Player*)ptr)->time;
}

int Player::curReefHealth()
{
  return ((_Player*)ptr)->curReefHealth;
}

int Player::sandDollars()
{
  return ((_Player*)ptr)->sandDollars;
}


bool Player::talk(char* message)
{
  return playerTalk( (_Player*)ptr, message);
}



std::ostream& operator<<(std::ostream& stream,Player ob)
{
  stream << "id: " << ((_Player*)ob.ptr)->id  <<'\n';
  stream << "playerName: " << ((_Player*)ob.ptr)->playerName  <<'\n';
  stream << "time: " << ((_Player*)ob.ptr)->time  <<'\n';
  stream << "curReefHealth: " << ((_Player*)ob.ptr)->curReefHealth  <<'\n';
  stream << "sandDollars: " << ((_Player*)ob.ptr)->sandDollars  <<'\n';
  return stream;
}
